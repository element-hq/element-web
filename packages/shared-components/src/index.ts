/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

// Components
export * from "./audio/AudioPlayerView";
export * from "./audio/Clock";
export * from "./audio/PlayPauseButton";
export * from "./audio/SeekBar";
export * from "./avatar/AvatarWithDetails";
export * from "./event-tiles/TextualEventView";
export * from "./message-body/MediaBody";
export * from "./pill-input/Pill";
export * from "./pill-input/PillInput";
export * from "./rich-list/RichItem";
export * from "./rich-list/RichList";
export * from "./utils/Box";
export * from "./utils/Flex";

// Utils
export * from "./utils/i18n"
export * from "./utils/humanize"
export * from "./utils/DateUtils"
export * from "./utils/numbers"

// MVVM
export * from "./ViewWrapper"
export * from "./ViewModel"
export * from "./useViewModel"
export * from "./MockViewModel"

