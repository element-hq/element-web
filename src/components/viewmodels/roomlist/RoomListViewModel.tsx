/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import type { Room } from "matrix-js-sdk/src/matrix";
import RoomListStoreV3 from "../../../stores/room-list-v3/RoomListStoreV3";

export interface RoomListViewState {
    /**
     * A list of rooms to be displayed in the left panel.
     */
    rooms: Room[];
}

/**
 * View model for the new room list
 * @see {@link RoomListViewState} for more information about what this view model returns.
 */
export function useRoomListViewModel(): RoomListViewState {
    const rooms = RoomListStoreV3.instance.getSortedRooms();
    return { rooms };
}
