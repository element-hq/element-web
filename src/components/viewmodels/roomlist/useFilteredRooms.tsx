/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { useCallback, useMemo, useState } from "react";

import type { Room } from "matrix-js-sdk/src/matrix";
import { FilterKey } from "../../../stores/room-list-v3/skip-list/filters";
import { _t, _td, type TranslationKey } from "../../../languageHandler";
import RoomListStoreV3 from "../../../stores/room-list-v3/RoomListStoreV3";
import { LISTS_UPDATE_EVENT } from "../../../stores/room-list/RoomListStore";
import { useEventEmitter } from "../../../hooks/useEventEmitter";
import SpaceStore from "../../../stores/spaces/SpaceStore";
import { UPDATE_SELECTED_SPACE } from "../../../stores/spaces";

/**
 * Provides information about a primary filter.
 * A primary filter is a commonly used filter that is given
 * more precedence in the UI. For eg, primary filters may be
 * rendered as pills above the room list.
 */
export interface PrimaryFilter {
    // A function to toggle this filter on and off.
    toggle: () => void;
    // Whether this filter is currently applied
    active: boolean;
    // Text that can be used in the UI to represent this filter.
    name: string;
    // The key of the filter
    key: FilterKey;
}

interface FilteredRooms {
    primaryFilters: PrimaryFilter[];
    rooms: Room[];
    activateSecondaryFilter: (filter: SecondaryFilters) => void;
    activeSecondaryFilter: SecondaryFilters;
    /**
     * The currently active primary filter.
     * If no primary filter is active, this will be undefined.
     */
    activePrimaryFilter?: PrimaryFilter;
}

const filterKeyToNameMap: Map<FilterKey, TranslationKey> = new Map([
    [FilterKey.UnreadFilter, _td("room_list|filters|unread")],
    [FilterKey.FavouriteFilter, _td("room_list|filters|favourite")],
    [FilterKey.PeopleFilter, _td("room_list|filters|people")],
    [FilterKey.RoomsFilter, _td("room_list|filters|rooms")],
]);

/**
 * These are the secondary filters which are not prominently shown
 * in the UI.
 */
export const enum SecondaryFilters {
    AllActivity,
    MentionsOnly,
    InvitesOnly,
    LowPriority,
}

/**
 * A map from {@link SecondaryFilters} which the UI understands to
 * {@link FilterKey} which the store understands.
 */
const secondaryFiltersToFilterKeyMap = new Map([
    [SecondaryFilters.AllActivity, undefined],
    [SecondaryFilters.MentionsOnly, FilterKey.MentionsFilter],
    [SecondaryFilters.InvitesOnly, FilterKey.InvitesFilter],
    [SecondaryFilters.LowPriority, FilterKey.LowPriorityFilter],
]);

/**
 * Use this function to determine if a given primary filter is compatible with
 * a given secondary filter. Practically, this determines whether it makes sense
 * to expose two filters together in the UI - for eg, it does not make sense to show the
 * favourite primary filter if the active secondary filter is low priority.
 * @param primary Primary filter key
 * @param secondary Secondary filter key
 * @returns true if compatible, false otherwise
 */
function isPrimaryFilterCompatible(primary: FilterKey, secondary: FilterKey): boolean {
    if (secondary === FilterKey.MentionsFilter) {
        if (primary === FilterKey.UnreadFilter) return false;
    } else if (secondary === FilterKey.InvitesFilter) {
        if (primary === FilterKey.UnreadFilter || primary === FilterKey.FavouriteFilter) return false;
    } else if (secondary === FilterKey.LowPriorityFilter) {
        if (primary === FilterKey.FavouriteFilter) return false;
    }
    return true;
}

/**
 * Track available filters and provide a filtered list of rooms.
 */
export function useFilteredRooms(): FilteredRooms {
    /**
     * Primary filter refers to the pill based filters
     * rendered above the room list.
     */
    const [primaryFilter, setPrimaryFilter] = useState<FilterKey | undefined>();
    /**
     * Secondary filters are also filters but they are hidden
     * away in a popup menu.
     */
    const [activeSecondaryFilter, setActiveSecondaryFilter] = useState<SecondaryFilters>(SecondaryFilters.AllActivity);

    const secondaryFilter = useMemo(
        () => secondaryFiltersToFilterKeyMap.get(activeSecondaryFilter),
        [activeSecondaryFilter],
    );

    const [rooms, setRooms] = useState(() => RoomListStoreV3.instance.getSortedRoomsInActiveSpace());

    const updateRoomsFromStore = useCallback((filters: FilterKey[] = []): void => {
        const newRooms = RoomListStoreV3.instance.getSortedRoomsInActiveSpace(filters);
        setRooms(newRooms);
    }, []);

    // Reset filters when active space changes
    useEventEmitter(SpaceStore.instance, UPDATE_SELECTED_SPACE, () => {
        setPrimaryFilter(undefined);
        activateSecondaryFilter(SecondaryFilters.AllActivity);
    });

    const filterUndefined = (array: (FilterKey | undefined)[]): FilterKey[] =>
        array.filter((f) => f !== undefined) as FilterKey[];

    const getAppliedFilters = (): FilterKey[] => {
        return filterUndefined([primaryFilter, secondaryFilter]);
    };

    useEventEmitter(RoomListStoreV3.instance, LISTS_UPDATE_EVENT, () => {
        const filters = getAppliedFilters();
        updateRoomsFromStore(filters);
    });

    /**
     * Secondary filters are activated using this function.
     * This is different to how primary filters work because the secondary
     * filters are static i.e they are always available and don't need to be
     * hidden.
     */
    const activateSecondaryFilter = useCallback(
        (filter: SecondaryFilters): void => {
            // If the filter is already active, just return.
            if (filter === activeSecondaryFilter) return;

            // SecondaryFilter is an enum for the UI, let's convert it to something
            // that the store will understand.
            const secondary = secondaryFiltersToFilterKeyMap.get(filter);
            setActiveSecondaryFilter(filter);

            // Reset any active primary filters.
            setPrimaryFilter(undefined);

            updateRoomsFromStore(filterUndefined([secondary]));
        },
        [activeSecondaryFilter, updateRoomsFromStore],
    );

    /**
     * This tells the view which primary filters are available, how to toggle them
     * and whether a given primary filter is active. @see {@link PrimaryFilter}
     */
    const primaryFilters = useMemo(() => {
        const createPrimaryFilter = (key: FilterKey, name: string): PrimaryFilter => {
            return {
                toggle: () => {
                    setPrimaryFilter((currentFilter) => {
                        const filter = currentFilter === key ? undefined : key;
                        updateRoomsFromStore(filterUndefined([filter, secondaryFilter]));
                        return filter;
                    });
                },
                active: primaryFilter === key,
                name,
                key,
            };
        };
        const filters: PrimaryFilter[] = [];
        for (const [key, name] of filterKeyToNameMap.entries()) {
            if (secondaryFilter && !isPrimaryFilterCompatible(key, secondaryFilter)) {
                continue;
            }
            filters.push(createPrimaryFilter(key, _t(name)));
        }
        return filters;
    }, [primaryFilter, updateRoomsFromStore, secondaryFilter]);

    const activePrimaryFilter = useMemo(() => primaryFilters.find((filter) => filter.active), [primaryFilters]);

    return { primaryFilters, activePrimaryFilter, rooms, activateSecondaryFilter, activeSecondaryFilter };
}
