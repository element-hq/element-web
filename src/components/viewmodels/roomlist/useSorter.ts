/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/
import { useState } from "react";

import RoomListStoreV3 from "../../../stores/room-list-v3/RoomListStoreV3";
import { SortingAlgorithm } from "../../../stores/room-list-v3/skip-list/sorters";
import SettingsStore from "../../../settings/SettingsStore";

/**
 * Sorting options made available to the view.
 */
export const enum SortOption {
    Activity = SortingAlgorithm.Recency,
    AToZ = SortingAlgorithm.Alphabetic,
}

/**
 * {@link SortOption} holds almost the same information as
 * {@link SortingAlgorithm}. This is done intentionally to
 * prevent the view from having a dependence on the
 * model (which is the store in this case).
 */
const sortingAlgorithmToSortingOption = {
    [SortingAlgorithm.Alphabetic]: SortOption.AToZ,
    [SortingAlgorithm.Recency]: SortOption.Activity,
};

const sortOptionToSortingAlgorithm = {
    [SortOption.AToZ]: SortingAlgorithm.Alphabetic,
    [SortOption.Activity]: SortingAlgorithm.Recency,
};

interface SortState {
    sort: (option: SortOption) => void;
    activeSortOption: SortOption;
}

/**
 * This hook does two things:
 * - Provides a way to track the currently active sort option.
 * - Provides a function to resort the room list.
 */
export function useSorter(): SortState {
    const [activeSortingAlgorithm, setActiveSortingAlgorithm] = useState(() =>
        SettingsStore.getValue("RoomList.preferredSorting"),
    );

    const sort = (option: SortOption): void => {
        const sortingAlgorithm = sortOptionToSortingAlgorithm[option];
        RoomListStoreV3.instance.resort(sortingAlgorithm);
        setActiveSortingAlgorithm(sortingAlgorithm);
    };

    return {
        sort,
        activeSortOption: sortingAlgorithmToSortingOption[activeSortingAlgorithm!],
    };
}
