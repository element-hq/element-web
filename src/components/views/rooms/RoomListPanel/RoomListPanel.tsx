/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";

import { shouldShowComponent } from "../../../../customisations/helpers/UIComponents";
import { UIComponent } from "../../../../settings/UIFeature";
import { RoomListSearch } from "./RoomListSearch";
import { RoomListHeaderView } from "./RoomListHeaderView";

type RoomListPanelProps = {
    /**
     * Current active space
     * See {@link RoomListSearch}
     */
    activeSpace: string;
};

/**
 * The panel of the room list
 */
export const RoomListPanel: React.FC<RoomListPanelProps> = ({ activeSpace }) => {
    const displayRoomSearch = shouldShowComponent(UIComponent.FilterContainer);

    return (
        <section className="mx_RoomListPanel" data-testid="room-list-panel">
            {displayRoomSearch && <RoomListSearch activeSpace={activeSpace} />}
            <RoomListHeaderView />
        </section>
    );
};
