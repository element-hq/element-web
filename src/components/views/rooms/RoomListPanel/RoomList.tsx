/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import React, { useCallback, type JSX } from "react";
import { AutoSizer, List, type ListRowProps } from "react-virtualized";

import { type RoomListViewState } from "../../../viewmodels/roomlist/RoomListViewModel";
import { _t } from "../../../../languageHandler";
import { RoomListItemView } from "./RoomListItemView";

interface RoomListProps {
    /**
     * The view model state for the room list.
     */
    vm: RoomListViewState;
}

/**
 * A virtualized list of rooms.
 */
export function RoomList({ vm: { rooms, activeIndex } }: RoomListProps): JSX.Element {
    const roomRendererMemoized = useCallback(
        ({ key, index, style }: ListRowProps) => (
            <RoomListItemView room={rooms[index]} key={key} style={style} isSelected={activeIndex === index} />
        ),
        [rooms, activeIndex],
    );

    // The first div is needed to make the virtualized list take all the remaining space and scroll correctly
    return (
        <div className="mx_RoomList" data-testid="room-list">
            <AutoSizer>
                {({ height, width }) => (
                    <List
                        aria-label={_t("room_list|list_title")}
                        className="mx_RoomList_List"
                        rowRenderer={roomRendererMemoized}
                        rowCount={rooms.length}
                        rowHeight={48}
                        height={height}
                        width={width}
                        scrollToIndex={activeIndex ?? 0}
                    />
                )}
            </AutoSizer>
        </div>
    );
}
