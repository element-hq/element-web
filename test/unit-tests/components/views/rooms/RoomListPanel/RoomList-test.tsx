/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import React from "react";
import { type MatrixClient } from "matrix-js-sdk/src/matrix";
import { render } from "jest-matrix-react";

import { mkRoom, stubClient } from "../../../../../test-utils";
import { type RoomListViewState } from "../../../../../../src/components/viewmodels/roomlist/RoomListViewModel";
import { RoomList } from "../../../../../../src/components/views/rooms/RoomListPanel/RoomList";
import DMRoomMap from "../../../../../../src/utils/DMRoomMap";
import { SecondaryFilters } from "../../../../../../src/components/viewmodels/roomlist/useFilteredRooms";
import { SortOption } from "../../../../../../src/components/viewmodels/roomlist/useSorter";

describe("<RoomList />", () => {
    let matrixClient: MatrixClient;
    let vm: RoomListViewState;

    beforeEach(() => {
        // Needed to render the virtualized list in rtl tests
        // https://github.com/bvaughn/react-virtualized/issues/493#issuecomment-640084107
        jest.spyOn(HTMLElement.prototype, "offsetHeight", "get").mockReturnValue(1500);
        jest.spyOn(HTMLElement.prototype, "offsetWidth", "get").mockReturnValue(1500);

        matrixClient = stubClient();
        const rooms = Array.from({ length: 10 }, (_, i) => mkRoom(matrixClient, `room${i}`));
        vm = {
            rooms,
            primaryFilters: [],
            activateSecondaryFilter: () => {},
            activeSecondaryFilter: SecondaryFilters.AllActivity,
            sort: jest.fn(),
            activeSortOption: SortOption.Activity,
            shouldShowMessagePreview: false,
            toggleMessagePreview: jest.fn(),
            createRoom: jest.fn(),
            createChatRoom: jest.fn(),
            canCreateRoom: true,
            activeIndex: undefined,
        };

        // Needed to render a room list cell
        DMRoomMap.makeShared(matrixClient);
        jest.spyOn(DMRoomMap.shared(), "getUserIdForRoomId").mockReturnValue(null);
    });

    it("should render a room list", () => {
        const { asFragment } = render(<RoomList vm={vm} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
