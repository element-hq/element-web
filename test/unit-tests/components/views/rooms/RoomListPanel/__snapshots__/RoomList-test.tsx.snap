// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomList /> should render a room list 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomList"
    data-testid="room-list"
  >
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    >
      <div
        aria-label="Room list"
        aria-readonly="true"
        class="ReactVirtualized__Grid ReactVirtualized__List mx_RoomList_List"
        role="grid"
        style="box-sizing: border-box; direction: ltr; height: 1500px; position: relative; width: 1500px; will-change: transform; overflow-x: hidden; overflow-y: hidden;"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Grid__innerScrollContainer"
          role="row"
          style="width: auto; height: 480px; max-width: 1500px; max-height: 480px; overflow: hidden; position: relative;"
        >
          <button
            aria-label="Open room room0"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 0px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="2"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room0"
                >
                  room0
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room1"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 48px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="3"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room1"
                >
                  room1
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room2"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 96px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="4"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room2"
                >
                  room2
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room3"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 144px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="5"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room3"
                >
                  room3
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room4"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 192px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="6"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room4"
                >
                  room4
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room5"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 240px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="1"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room5"
                >
                  room5
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room6"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 288px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="2"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room6"
                >
                  room6
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room7"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 336px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="3"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room7"
                >
                  room7
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room8"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 384px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="4"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room8"
                >
                  room8
                </span>
              </div>
            </div>
          </button>
          <button
            aria-label="Open room room9"
            aria-selected="false"
            class="mx_RoomListItemView mx_RoomListItemView_empty"
            role="gridcell"
            style="height: 48px; left: 0px; position: absolute; top: 432px; width: 100%;"
            type="button"
          >
            <div
              class="mx_Flex mx_RoomListItemView_container"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
            >
              <div
                class="mx_DecoratedRoomAvatar"
              >
                <span
                  aria-label="Avatar"
                  class="_avatar_1qbcf_8 mx_BaseAvatar"
                  data-color="5"
                  data-testid="avatar-img"
                  data-type="round"
                  style="--cpd-avatar-size: 32px;"
                >
                  <img
                    alt=""
                    class="_image_1qbcf_41"
                    data-type="round"
                    height="32px"
                    loading="lazy"
                    referrerpolicy="no-referrer"
                    src="http://this.is.a.url/avatar.url/room.png"
                    width="32px"
                  />
                </span>
              </div>
              <div
                class="mx_Flex mx_RoomListItemView_content"
                style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x); --mx-flex-wrap: nowrap;"
              >
                <span
                  class="mx_RoomListItemView_roomName"
                  title="room9"
                >
                  room9
                </span>
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1501px; height: 1501px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</DocumentFragment>
`;
